### ISSUE TITLE
Move tab to a new window

### ISSUE THREAD

**Issue by mitchellh** [ghostty-org/ghostty#2630]
### Discussed in https://github.com/ghostty-org/ghostty/discussions/2625

<div type='discussions-op-text'>

<sup>Originally posted by **jcollie** November  8, 2024</sup>
Currently (with libadwaita enabled) you can move a tab to another existing window by dragging the tab header from one window to another. I propose adding a keybind action and corresponding context menu entries to move a tab to a new (previously non-existent) window where it would become the first and only tab in that window.</div>

### Design

1. Add `move_tab_to_new_window` binding
2. Add the associated apprt action
3. Implement for GTK and/or macOS
We can support both platforms for this one.

**Comment by kvz** [comment]
> Currently (with libadwaita enabled) you can move a tab to another existing window by dragging the tab header from one window to another

Just adding a datapoint, I could not get this to work on Omarchy/Hyprland. The tab will just immediately snap back to its original window after dragging it to another.

**Comment by gwuen** [comment]
> Just adding a datapoint, I could not get this to work on Omarchy/Hyprland. The tab will just immediately snap back to its original window after dragging it to another.

Moving tabs only seems works between windows of the same instance. Depending on the way you start Ghostty, you might create new instances without realizing. This behavior is configurable with the `gtk-single-instance` option, with the default value being `detect`:

> If detect, Ghostty will assume true (single instance) unless one of the following scenarios is found:
>
> 1. TERM_PROGRAM environment variable is a non-empty value. In this case, we assume Ghostty is being launched from a graphical terminal session and you want a dedicated instance.
>
> 2. Any CLI arguments exist. In this case, we assume you are passing custom Ghostty configuration. Single instance mode inherits the configuration from when it was launched, so we must disable single instance to load the new configuration.

Moving a single tab to a *new window* should be unaffected though. This seems to be fiddly in general, but dragging it to the edge of the screen generally works for me.

**Comment by jcollie** [comment]
> > Just adding a datapoint, I could not get this to work on Omarchy/Hyprland. The tab will just immediately snap back to its original window after dragging it to another.
> 
> Moving tabs only seems works between windows of the same instance. Depending on the way you start Ghostty, you might create new instances without realizing. This behavior is configurable with the `gtk-single-instance` option, with the default value being `detect`:

Yes, this is a fundamental limitation as moving tabs between windows (or to a new window) requires access to the shared data structures in memory, both in the Ghostty code as well as the data structures maintained by GTK. Passing a window to a different process would require a complicated protocol to synchronize the data structures as well as passing the PTY between processes. Not necessarily _impossible_ but definitely more work than it's worth.

> Moving a single tab to a _new window_ should be unaffected though. This seems to be fiddly in general, but dragging it to the edge of the screen generally works for me.

Hyprland probably makes that more difficult as it's generally completely covering the screen with other windows.

**Comment by joedevon** [comment]
> > Just adding a datapoint, I could not get this to work on Omarchy/Hyprland. The tab will just immediately snap back to its original window after dragging it to another.
> 
> Moving tabs only seems works between windows of the same instance. Depending on the way you start Ghostty, you might create new instances without realizing. This behavior is configurable with the `gtk-single-instance` option, with the default value being `detect`:
> 
> > If detect, Ghostty will assume true (single instance) unless one of the following scenarios is found:
> > 
> > 1. TERM_PROGRAM environment variable is a non-empty value. In this case, we assume Ghostty is being launched from a graphical terminal session and you want a dedicated instance.
> > 2. Any CLI arguments exist. In this case, we assume you are passing custom Ghostty configuration. Single instance mode inherits the configuration from when it was launched, so we must disable single instance to load the new configuration.
> 
> Moving a single tab to a _new window_ should be unaffected though. This seems to be fiddly in general, but dragging it to the edge of the screen generally works for me.

I've been trying to get a tab to attach to a window in the same instance but it doesn't work for me on mac. Has anyone gotten it to work?

**Comment by foucist** [comment]
In Omarchy/Hyperland, SUPER+ENTER opens a Ghostty window by default.  If I start some tabs inside it, and then I want to move one of the tabs out into a separate window, I apparently have to open another Ghostty window via SUPER+ENTER and then I can drag the tab over.  Unfortunately this leaves me with a 2-tabbed Ghostty window, which feels messy and not quite what I wanted.

I'd love for a keybinding action that can essentially spawn a Ghostty window, and move the focused tab over in one go. 

Is this possible?

**Comment by Sxderp** [comment]
I would also like to request support for this. I'm surprised it wasn't already a feature..

**Comment by Ethannewer** [comment]
Really looking forward to this feature! Being able to detach a tab into its own window via keyboard shortcut would be a huge productivity boost. It's the one thing I miss compared to other terminals — sometimes you just want to pull a session out into its own window without reaching for the mouse and dragging. Hoping this lands soon, keep up the great work! 🚀

**Comment by globalhuman** [comment]
I was using iterm2 and vscode for terminal management before switching to ghostty and really enjoyed the endless customizability for window/pane management. I do hope that this gets added soon.

macOS user.

**Comment by hdprajwal** [comment]
A workaround that i use is to open a second ghostty window with `SUPER+ENTER` and drag the tab that i want into the new window then close the extra tab that was created. this works but then it's a lot more steps. 
It would be such a timesaver if we can do this with a single key bind or drag and drop a tab into its own new window.

**Comment by SanJJ1** [comment]
I have a working implementation of this per the design notes above (macOS only: keybind action + command palette entry), built and manually tested on macOS 26. Vouch request pending: https://github.com/ghostty-org/ghostty/discussions/13436 
                                                                                                                                                                                                                                                
Posting for visibility so no one duplicates the work; I'll open a PR once vouched.

**Comment by pedronaugusto** [comment]
I've got the macOS side of this, implemented and pushed: https://github.com/pedronaugusto/ghostty/tree/move-tab-new-window-macos. One commit, `move_tab_to_new_window` forwards to AppKit's `moveTabToNewWindow:`. Tested on macOS 26 from a keybind and from the command palette. I'll open the PR once I'm vouched (#14166).

@SanJJ1 you mentioned an implementation in July. Are you still opening a PR? I'll hold off if you are.
